### PR TITLE
[CI] Build wheels on push

### PR DESCRIPTION
**Pull request by phschiele**
## Description
Our current build process only builds wheels when we release.
Often, failures of these builds are unrelated to the code changes. When wheels are partially released to PyPI, we have to push a fix in a subsequent release. 

This change builds whenever we merge to master or a release branch (but not on every PR commit).

Regarding changes to the procedures, @h-vetinari you [mentioned](https://github.com/cvxpy/cvxpy/pull/1998#issuecomment-1370662440) that its possible to test pre-releases against your infrastructure. Could you create an example PR on the cvxpy feedstock repo that shows how this is done, then I could add a reference to that in the PROCEDURES document.

## Type of change
- [ ] New feature (backwards compatible)
- [ ] New feature (breaking API changes)
- [ ] Bug fix
- [x] Other (Documentation, CI, ...)

## [Contribution checklist](https://www.cvxpy.org/contributing/index.html#contribution-checklist)
- [ ] Add our license to new files.
- [ ] Check that your code adheres to our coding style.
- [ ] Write unittests.
- [ ] Run the unittests and check that they’re passing.
- [ ] Run the benchmarks to make sure your change doesn’t introduce a regression.

### CODE REVIEW

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -5,17 +5,17 @@ on:
...
+  cancel-in-progress: true
+
 jobs:
-  cleanup-runs:
-    runs-on: ubuntu-latest
-    steps:
-    - uses: rokroskar/workflow-run-cleanup-action@master
-      env:
-        GITHUB_TOKEN: "${{ secrets.GITHUB_TOKEN }}"
-    if: "!startsWith(github.ref, 'refs/tags/') && github.ref != 'refs/heads/master'"
```
Comment (phschiele) [inline .github/workflows/build.yml:18]: This action is deprecated and replaced by the builtin `concurrency` keyword now.

File: .github/workflows/build.yml [hunk 2]
```diff
@@ -5,17 +5,17 @@ on:
     push:
         branches:
             - master
+            - 'release/**'
         tags:
           - '*'
 
+# https://docs.github.com/en/enterprise-cloud@latest/actions/using-jobs/using-concurrency#example-using-a-fallback-value
```
Comment (phschiele) [inline .github/workflows/build.yml:12]:
```suggestion
# https://docs.github.com/en/actions/using-jobs/using-concurrency#example-using-a-fallback-value
```

**Comment by h-vetinari** [comment]
> Regarding changes to the procedures, @h-vetinari you [mentioned](https://github.com/cvxpy/cvxpy/pull/1998#issuecomment-1370662440) that its possible to test pre-releases against your infrastructure. Could you create an example PR on the cvxpy feedstock repo that shows how this is done, then I could add a reference to that in the PROCEDURES document.

While it's possible to point to an arbitrary commit, the easiest is to have a tagged (pre-)release, and then testing is as simple as pointing the conda-forge infrastructure to that tag - as a recent example from another feedstock, see [here](https://github.com/conda-forge/openssl-feedstock/pull/116/commits/995e15fa86201fe50f631153b04902e4f9e91158).

File: .github/workflows/build.yml [hunk 3]
```diff
@@ -125,28 +125,33 @@ jobs:
...
+
+      - name:  Release to pypi
+        if: ${{env.DEPLOY == 'True' &&  env.USE_OPENMP != 'True'}}
+        shell: bash
+        run: |
           twine upload --skip-existing --repository-url $PYPI_SERVER wheelhouse/* -u $PYPI_USER -p $PYPI_PASSWORD
 
       - name: Upload artifacts to github
-        if: ${{env.DEPLOY == 'True' && env.USE_OPENMP != 'True'}}
+        if: ${{github.event_name == 'push' && env.USE_OPENMP != 'True'}}
```
Comment (SteveDiamond) [inline .github/workflows/build.yml:154]: Shouldn't this have the same conditions as release to pypi?

File: .github/workflows/build.yml [hunk 4]
```diff
@@ -209,28 +214,33 @@ jobs:
...
+        if: ${{github.event_name == 'push'}}
         run: |
           python -m pip install --upgrade twine
           twine check wheelhouse/*
+
+      - name: Release to pypi
+        shell: bash
+        if: ${{env.DEPLOY == 'True'}}
+        run: |
           twine upload --skip-existing --repository-url $PYPI_SERVER wheelhouse/* -u $PYPI_USER -p $PYPI_PASSWORD
```
Comment (SteveDiamond) [inline .github/workflows/build.yml:240]: Change $PYPI_PASSWORD to $PYPI_BASE_PASSWORD

File: .github/workflows/build.yml (hunk 3 above)
Comment (phschiele) [reply .github/workflows/build.yml:154]: This is just uploading to the artifacts section of the run on GitHub, so we can try out the wheels generated by the run if required.
It does **not** generate a release or similar.

File: .github/workflows/build.yml (hunk 4 above)
Comment (phschiele) [reply .github/workflows/build.yml]: Done in [fe338b9](https://github.com/cvxpy/cvxpy/pull/2008/commits/fe338b9ffa9a1eaf61cd954e8a750c0be78c50a2)

**Review by SteveDiamond — APPROVED** [review]

**Pull request merged**